src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -200,24 +207,72 @@ RouterExecutorRun(QueryDesc *queryDesc, ScanDirection direction, long count, Tas
 		InstrStartNode(queryDesc->totaltime);
 	}
 
-	if (operation == CMD_INSERT || operation == CMD_UPDATE ||
-		operation == CMD_DELETE)
+	estate->es_processed = 0;
+
+	/* startup the tuple receiver */
+	if (sendTuples)
 	{
-		int32 affectedRowCount = ExecuteDistributedModify(task);
-		estate->es_processed = affectedRowCount;
+		(*destination->rStartup)(destination, operation, queryDesc->tupDesc);
 	}
-	else if (operation == CMD_SELECT)
+
+	/*
+	 * If query has not yet been executed, do so now. The main reason why the
+	 * query might already have been executed is cursors.
+	 */
+	if (!routerState->eof_underlying)
+	{
+		bool resultsOK = false;
+		bool isModificationQuery = false;
```

Comment:
> >  	}
> > -	else if (operation == CMD_SELECT)
> > +
> > +	/*
> > +	 * If query has not yet been executed, do so now. The main reason why the
> > +	 * query might already have been executed is cursors.
> > +	 */
> > +	if (!routerState->eof_underlying)
> > +	{
> > +		bool resultsOK = false;
> > +		bool isModificationQuery = false;
> 
> Minor again, the coding convention document suggests to avoid `Is`, instead use `modificationQuery`

I think that's bad here. modificationQuery sounds like it'd contain an
actual Query.